packet_manager.go

```
@@ -0,0 +1,43 @@
+package nex
+
+// PacketManager implements an API for pushing/popping packets in the correct order
+type PacketManager struct {
+	currentSequenceID *Counter
+	packets           []PacketInterface
+}
+
+// Next gets the next packet in the sequence. Returns nil if the next packet has not been sent yet
+func (p *PacketManager) Next() PacketInterface {
+	var packet PacketInterface
+
+	for i := 0; i < len(p.packets); i++ {
+		if p.currentSequenceID.Value() == uint32(p.packets[i].SequenceID()) {
+			packet = p.packets[i]
+			p.RemoveByIndex(i)
+			p.currentSequenceID.Increment()
+			break
+		}
+	}
+
+	return packet
+}
+
+// Push adds a packet to the pool to choose from in Next
+func (p *PacketManager) Push(packet PacketInterface) {
```

Shouldn't this be checking for duplicated packets?

---

A duplicate packet will always have a sequence ID less than the currently expected one. The `Next()` method already ignores these kinds of packets, since it's only checking for the next sequence ID, so them being in the pool doesn't really cause any issues. Even in cases where theres multiple of the same packet in the pool at the same time, `Next()` will only pick the 1st one and then increment, meaning any duplicates never get touched. Since this does not happen that often, it would be generally more wasteful to store every seen packet ID to explicitly check for duplicates, compared to only storing a (comparatively) small number of stale pointers (which will get cleaned up by the GC eventually anyway)

Even though it shouldn't cause any issues the way it is, it would probably be a good little micro-optimization to just not append to the slice in these cases. Though I don't think it should error, just silently ignore?